// styled-components-carousel/dist/Carousel.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var styled_components_1 = __importDefault(require("styled-components"));
var Slider_1 = __importDefault(require("./Slider"));
var Indicator_1 = __importDefault(require("./Indicator"));
var Arrow_1 = __importStar(require("./Arrow"));
var Swipeable_1 = require("./Swipeable");
var helpers_1 = require("./helpers");
var CarouselWrapper = styled_components_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n    display: flex;\n    flex-direction: column;\n    justify-content: center;\n    align-items: center;\n    position: relative;\n    overflow: hidden;\n    width: 100%;\n"], ["\n    display: flex;\n    flex-direction: column;\n    justify-content: center;\n    align-items: center;\n    position: relative;\n    overflow: hidden;\n    width: 100%;\n"])));
var SliderWrapper = styled_components_1.default.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n    display: flex;\n    flex-direction: row;\n    align-items: center;\n    width: 100%;\n"], ["\n    display: flex;\n    flex-direction: row;\n    align-items: center;\n    width: 100%;\n"])));
var Component = function (_a) {
    var _b = _a.showIndicator, showIndicator = _b === void 0 ? true : _b, _c = _a.showArrows, showArrows = _c === void 0 ? true : _c, _d = _a.swipeable, swipeable = _d === void 0 ? true : _d, _e = _a.slidesToShow, slidesToShow = _e === void 0 ? 1 : _e, _f = _a.centerPadding, centerPadding = _f === void 0 ? 0 : _f, _g = _a.scaleOnFocus, scaleOnFocus = _g === void 0 ? 1 : _g, breakpoints = _a.breakpoints, center = _a.center, infinite = _a.infinite, debug = _a.debug, children = _a.children;
    var childrenCount = react_1.default.Children.count(children);
    var carouselSettings = {
        center: center,
        infinite: infinite,
        slidesToShow: slidesToShow,
        showArrows: showArrows,
        showIndicator: showIndicator,
        swipeable: swipeable,
        centerPadding: centerPadding,
        scaleOnFocus: scaleOnFocus,
    };
    var _h = react_1.useState(helpers_1.matchBreakpoint(window.innerWidth, carouselSettings, breakpoints)), activeSettings = _h[0], setActiveSettings = _h[1];
    var _j = react_1.useState({
        previousActive: 0,
        active: 0,
        infiniteActive: 0,
    }), _k = _j[0], previousActive = _k.previousActive, active = _k.active, infiniteActive = _k.infiniteActive, setActive = _j[1];
    return (react_1.default.createElement(CarouselWrapper, null,
        react_1.default.createElement(SliderWrapper, null,
            activeSettings.showArrows && (react_1.default.createElement(Arrow_1.default, { type: Arrow_1.ArrowType.previous, onClick: function () { return setActive(helpers_1.getIndexForAction(active, active - 1, childrenCount)); }, disabled: !helpers_1.canGoPrevious(active, activeSettings.infinite) })),
            react_1.default.createElement(Slider_1.default, __assign({ previousActive: previousActive, active: active, infiniteActive: infiniteActive, onWindowResize: function () {
                    return setActiveSettings(helpers_1.matchBreakpoint(window.innerWidth, carouselSettings, breakpoints));
                }, debug: debug, onSwipe: function (direction) {
                    var newActive = active;
                    if (direction === Swipeable_1.SwipeDirection.Left &&
                        helpers_1.canGoNext(active, childrenCount, activeSettings.slidesToShow || slidesToShow, activeSettings.infinite)) {
                        newActive = active + 1;
                    }
                    else if (direction === Swipeable_1.SwipeDirection.Right &&
                        helpers_1.canGoPrevious(active, activeSettings.infinite)) {
                        newActive = active - 1;
                    }
                    setActive(helpers_1.getIndexForAction(active, newActive, childrenCount));
                } }, activeSettings), children),
            activeSettings.showArrows && (react_1.default.createElement(Arrow_1.default, { type: Arrow_1.ArrowType.next, onClick: function () { return setActive(helpers_1.getIndexForAction(active, active + 1, childrenCount)); }, disabled: !helpers_1.canGoNext(active, childrenCount, activeSettings.slidesToShow || slidesToShow, activeSettings.infinite) }))),
        activeSettings.showIndicator && (react_1.default.createElement(Indicator_1.default, { items: childrenCount, slidesToShow: activeSettings.slidesToShow || slidesToShow, infinite: activeSettings.infinite, active: infiniteActive, onClick: function (index) { return setActive(helpers_1.getIndexForAction(active, index, childrenCount)); } }))));
};
exports.default = Component;
var templateObject_1, templateObject_2;
