// react-easy-marquee/dist/index.es.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { useState } from 'react';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

var Animation = function (_a) {
    var axis = _a.axis, reverse = _a.reverse, offset = _a.offset;
    return (jsx("style", { children: "\n        @keyframes slide" + offset + " {\n          from {\n            transform: translate" + (axis || "X") + "(" + offset * 100 + "%);\n          }\n          to {\n            transform: translate" + (axis || "X") + "(" + ((reverse ? -100 : 100) + 100 * offset) + "%);\n          }\n        }\n      " }, void 0));
};

var Marquee = function (_a) {
    var axis = _a.axis, reverse = _a.reverse, align = _a.align, background = _a.background, duration = _a.duration, height = _a.height, pauseOnHover = _a.pauseOnHover, width = _a.width, className = _a.className, children = _a.children;
    var _b = useState("running"), animate = _b[0], setAnimate = _b[1];
    var offsetValues = [-1, 0, 1];
    return (jsx("div", __assign({ className: className, onMouseEnter: function () { return pauseOnHover && setAnimate("paused"); }, onMouseLeave: function () { return pauseOnHover && setAnimate("running"); }, style: {
            display: "flex",
            overflow: "hidden",
            alignItems: "center",
            position: "relative",
            backgroundColor: background || "transparent",
            height: height || "5rem",
            width: width || "100%",
        } }, { children: offsetValues.map(function (offset) { return (jsxs("section", __assign({ style: {
                whiteSpace: "nowrap",
                overflow: "hidden",
                position: "absolute",
                alignItems: "center",
                display: "flex",
                minWidth: "100%",
                minHeight: "100%",
            } }, { children: [jsx(Animation, { reverse: reverse, offset: offset, axis: axis }, void 0),
                jsx("div", __assign({ style: {
                        width: "100%",
                        height: "100%",
                        whiteSpace: "nowrap",
                        display: "flex",
                        overflow: "hidden",
                        justifyContent: "space-around",
                        animation: "slide" + offset + " " + (duration || 5000) + "ms linear infinite",
                        animationPlayState: animate || "running",
                        flexDirection: axis === "X"
                            ? reverse
                                ? "row-reverse"
                                : "row"
                            : axis === "Y"
                                ? reverse
                                    ? "column-reverse"
                                    : "column"
                                : "row",
                        alignItems: align === "end" || align === "start"
                            ? "flex-" + align
                            : "center",
                    } }, { children: children }), void 0)] }), offset)); }) }), void 0));
};

export default Marquee;
